pyatlan/model/packages/base/crawler.py

```diff
@@ -0,0 +1,116 @@
+from json import dumps
+from typing import Any, Dict, Optional
+
+from pyatlan import utils
+from pyatlan.cache.role_cache import RoleCache
+from pyatlan.errors import ErrorCode
+from pyatlan.model.assets import Connection
+from pyatlan.model.packages.base.package import AbstractPackage
+
+
+class AbstractCrawler(AbstractPackage):
+    """
+    Abstract class for crawlers
+
+    :param connection_name: name for the connection
+    :param connection_type: type of connector for the connection
+    :param admin_roles: admin roles for the connection
+    :param admin_groups: admin groups for the connection
+    :param admin_users: admin users for the connection
+    :param allow_query: allow data to be queried in the connection (True) or not (False)
+    :param allow_query_preview: allow sample data viewing for assets in the connection (True) or not (False)
+    :param row_limit: maximum number of rows that can be returned by a query
+    :param source_logo: logo to use for the source
+
+    :raises AtlanException: if there is not at least one role,
+    group, or user defined as an admin (or any of them are invalid)
+    """
+
+    def __init__(
+        self,
+        connection_name: str,
+        connection_type: str,
+        admin_roles: Optional[list[str]],
+        admin_groups: Optional[list[str]],
+        admin_users: Optional[list[str]],
+        allow_query: bool = False,
+        allow_query_preview: bool = False,
+        row_limit: int = 0,
+        source_logo: str = "",
+    ):
+        self._parameters: list = []
+        self._credentials_body: dict = {}
+        self._epoch = int(utils.get_epoch_timestamp())
+        self._connection_name = connection_name
+        self._connection_type = connection_type
+        self._admin_roles = admin_roles
+        self._admin_groups = admin_groups
+        self._admin_users = admin_users
+        self._allow_query = allow_query
+        self._allow_query_preview = allow_query_preview
+        self._row_limit = row_limit
+        self._source_logo = source_logo
+
+    def _get_connection(self) -> Connection:
+        """
+        Builds a connection using the provided parameters,
+        which will be the target for the package to crawl assets.
+        """
+        if not self._admin_roles:
+            role = RoleCache.get_id_for_name("$admin")
+            if role:
+                self._admin_roles = [role]
```

Comment:
This is tempting to use as a fallback, but possibly changes the intended behavior. It's perfectly permissible that someone will create a connection that only has connection admins as a particular set of users or a group — and will not allow any roles as connection admins.

In this case, this would set everyone with the Admin role (either currently or at some point in the future) as a connection admin, which would violate the desired behavior above.

If a user wants to set all Admins as connection admins, they therefore need to be explicit rather than us doing it automatically every time the provided list of admin roles is empty.

Reply:
Understood. I have removed that condition. Thank you! 👍